File: 8_kyu/Safen_User_Input_Part_I_htmlspecialchars.py
# The standard escape method in the html package does not fulfill the specification.
# Implemention based on a single regex replace function call.
import re

# Translation table of the specified special characters.
TRANSLATE = {
    "<" : "&lt;",
    ">" : "&gt;",
    '"' : "&quot;",
    "&" : "&amp;"}

# Pattern matching any special character.
RE_SPECIAL_CHAR = re.compile("|".join(re.escape(k) for k in TRANSLATE))

# Translation of a special character match.
def translate(m: re.Match) -> str:
    return TRANSLATE[m.group(0)]

def html_special_chars(data: str) -> str:
    return RE_SPECIAL_CHAR.sub(translate, data)

File: 8_kyu/Safen_User_Input_Part_I_htmlspecialchars.rs
// Return string with escaped HTML special characters.
fn html_special_chars(html: &str) -> String {
    let mut accu = String::with_capacity(html.len() * 2);
    for c in html.chars() {
        match c {
            '<' => accu.push_str("&lt;"),
            '>' => accu.push_str("&gt;"),
            '&' => accu.push_str("&amp;"),
            '"' => accu.push_str("&quot;"),
            _ => accu.push(c),
        }
    }
    accu
}
